core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -676,8 +678,19 @@ object SparkSubmit {
       addJarToClasspath(jar, loader)
     }
 
-    for ((key, value) <- sysProps) {
-      System.setProperty(key, value)
+    val threadEnabled = sysProps.getOrElse("spark.launcher.internal.launcher.thread.enabled",
+       "false").toBoolean
+    if (!threadEnabled) {
+      for ((key, value) <- sysProps) {
+        System.setProperty(key, value)
+      }
+    } else {
+      for ( (key, value) <- sys.env) {
```

Comment:
I commented on this before, but I think it's extremely weird to be mixing system properties and env variables. They really should be separate parameters to the `sparkMain` method.